File: src/main/java/javafx/Draggable.java
package javafx;

import application.Storage;
import javafx.collections.ObservableList;
import javafx.event.Event;
import javafx.event.EventHandler;
import javafx.geometry.Point2D;
import javafx.scene.input.MouseEvent;
import javafx.scene.layout.Pane;
import javafx.scene.shape.*;
import math.Intersection;
import optics.light.Ray;
import optics.light.RayCircle;
import optics.objects.OpticalRectangle;

import java.util.ArrayList;

public class Draggable {
  private Point2D movementDelta;
  private Shape shape;
  private EventHandler<Event> onDrag;
  private EventHandler<Event> onDestroy;
  private Pane parent;

  public Draggable(Shape s, EventHandler<Event> onDrag, EventHandler<Event> onDestroy, Pane parent) {
    this.shape = s;
    this.onDrag = onDrag;
    this.onDestroy = onDestroy;
    this.parent = parent;
    if (this.shape instanceof Rectangle) {
      this.shape.setOnMousePressed(event -> {
        movementDelta = new Point2D(((Rectangle) this.shape).getX() - event.getSceneX(), ((Rectangle) this.shape)
                .getY() - event.getSceneY());
        event.consume();
      });
    } else if (this.shape instanceof RayCircle) {
      this.shape.setOnMousePressed(event -> {
        movementDelta = new Point2D(((Circle) this.shape).getCenterX() - event.getSceneX(), ((Circle) this.shape)
                .getCenterY() - event.getSceneY());
        event.consume();
      });
    }

    shape.setOnMouseDragged(event -> {
      // Prevent changes when animating
      if (Storage.isAnimating) return;
      // Prevent object from entering UI area
      double prevY = 0;
      if (this.shape instanceof OpticalRectangle) {
        prevY = ((OpticalRectangle) this.shape).getY();
        ((OpticalRectangle) this.shape).setScreenX(event.getSceneX() + movementDelta.getX());
        ((OpticalRectangle) this.shape).setScreenY(event.getSceneY() + movementDelta.getY());
      } else if (this.shape instanceof Circle) {
        prevY = ((Circle) this.shape).getCenterY();
        Ray r = ((RayCircle) this.shape).getRay();
        r.setScreenX(event.getSceneX() + movementDelta.getX());
        r.setScreenY(event.getSceneY() + movementDelta.getY());
      }
      if (isInUIArea(event) &&
              //        Except when moving object to trash
              !(event.getSceneX() > (parent.getWidth() - 82))) {
        if (this.shape instanceof OpticalRectangle) ((OpticalRectangle) this.shape).setScreenY(prevY);
        else if (this.shape instanceof RayCircle) ((RayCircle) this.shape).getRay().setScreenY(prevY);
      }
      if (this.onDrag == null) {
        event.consume();
        return;
      }
      this.onDrag.handle(event);
      event.consume();
    });
    shape.setOnMouseReleased(e -> {
      if (e.getSceneY() > (parent.getHeight() - 165) && e.getSceneX() > (parent.getWidth() - 82)) {
        parent.getChildren().remove(this.shape);
        this.onDestroy.handle(e);
      }
    });
  }

  private boolean isInUIArea(MouseEvent event) {
    if (shape instanceof Circle) {
      return event.getSceneY() >= (parent.getHeight() - 165);
    } else if (shape instanceof Rectangle) {
      if (((Rectangle) shape).getY() + ((Rectangle) shape).getWidth() + ((Rectangle) shape).getHeight() < parent
              .getHeight() - 165) {
        return false;
      }
    }
    ObservableList<PathElement> intersection = ((Path) Shape.intersect(shape, shape)).getElements();
    ArrayList<Point2D> points = Intersection.convertToPoints(intersection);
    return (points.stream().anyMatch(point2D -> point2D.getY() >= (parent.getHeight() - 165)));
  }
}

File: src/main/java/javafx/Draggable.kt
package javafx

import application.Storage.reRenderAll
import javafx.event.EventHandler
import javafx.geometry.Point2D
import javafx.scene.input.MouseEvent
import javafx.scene.layout.Pane
import javafx.scene.shape.Circle
import javafx.scene.shape.Path
import javafx.scene.shape.Rectangle
import javafx.scene.shape.Shape
import math.Intersection

class Draggable(
    private val shape: Shape,
    private val parent: Pane,
    onDestroy: () -> Unit,
) {
    private var movementDelta: Point2D
    private fun isInUIArea(event: MouseEvent): Boolean {
        if (shape is Circle) {
            return event.sceneY >= parent.height - 165
        } else if (shape is Rectangle) {
            if (shape.y + shape.width + shape.height < parent
                    .height - 165
            ) {
                return false
            }
        }
        val intersection = (Shape.intersect(shape, shape) as Path).elements
        val points = Intersection.convertToPoints(intersection)
        return points.stream()
            .anyMatch { point2D: Point2D -> point2D.y >= parent.height - 165 }
    }

    init {
        movementDelta = Point2D.ZERO
//        if (shape is Rectangle) {
//            shape.setOnMousePressed { event: MouseEvent ->
//                movementDelta = Point2D(
//                    shape.x - event.sceneX, shape
//                        .y - event.sceneY)
//                event.consume()
//            }
//        } else if (shape is RayCircle) {
//            shape.setOnMousePressed { event: MouseEvent ->
//                movementDelta =
//                    Point2D((shape as Circle).centerX - event.sceneX,
//                        (shape as Circle)
//                            .centerY - event.sceneY)
//
//                event.consume()
//            }
//        }
        shape.onMouseDragged = EventHandler { event: MouseEvent ->
//            // Prevent changes when animating
//            if (Storage.isAnimating) return@EventHandler
//            // Prevent object from entering UI area
//            var prevY = 0.0
//            if (shape is OpticalRectangle) {
//                prevY = shape.y
//                shape.setScreenX(event.sceneX + movementDelta.x)
//                shape.setScreenY(event.sceneY + movementDelta.y)
//            } else if (shape is RayCircle) {
//                prevY = shape.centerY
//                val r = shape.ray
//                val newRealLine =
//                    r.realLine.copy(start = r.realLine.start - movementDelta,
//                        end = r.realLine.end - movementDelta)
//                r.update(newRealLine)
//                movementDelta =
//                    Point2D(shape.centerX - event.sceneX,
//                        shape.centerY - event.sceneY)
//                event.consume()
//                return@EventHandler
//            }
//            if (isInUIArea(event) &&  //        Except when moving object to trash
//                event.sceneX <= parent.width - 82
//            ) {
//                if (shape is OpticalRectangle) shape.setScreenY(prevY)
////                else if (shape is RayCircle) shape.ray.setScreenY(
////                    prevY)
//            }
//            event.consume()hen animating
//            if (Storage.isAnimating) return@EventHandler
//            // Prevent object from entering UI area
//            var prevY = 0.0
//            if (shape is OpticalRectangle) {
//                prevY = shape.y
//                shape.setScreenX(event.sceneX + movementDelta.x)
//                shape.setScreenY(event.sceneY + movementDelta.y)
//            } else if (shape is RayCircle) {
//                prevY = shape.centerY
//                val r = shape.ray
//                val newRealLine =
//                    r.realLine.copy(start = r.realLine.start - movementDelta,
//                        end = r.realLine.end - movementDelta)
//                r.update(newRealLine)
//                movementDelta =
//                    Point2D(shape.centerX - event.sceneX,
//                        shape.centerY - event.sceneY)
//                event.consume()
//                return@EventHandler
//            }
//            if (isInUIArea(event) &&  //        Except when moving object to trash
//                event.sceneX <= parent.width - 82
//            ) {
//                if (shape is OpticalRectangle) shape.setScreenY(prevY)
////                else if (shape is RayCircle) shape.ray.setScreenY(
////                    prevY)
//            }
//            event.consume()
            reRenderAll()
        }
        shape.onMouseReleased = EventHandler { e: MouseEvent ->
            if (e.sceneY > parent.height - 165 && e.sceneX > parent.width - 82) {
                parent.children.remove(shape)
                onDestroy()
            }
        }
    }
}
